src/librustdoc/core.rs

```diff
@@ -52,35 +50,35 @@ crate struct DocContext<'tcx> {
     /// Most of this logic is copied from rustc_lint::late.
     crate param_env: ParamEnv<'tcx>,
     /// Later on moved through `clean::Crate` into `cache`
-    crate external_traits: Rc<RefCell<FxHashMap<DefId, clean::TraitWithExtraInfo>>>,
+    crate external_traits: Rc<RefCell<FxHashMap<RealDefId, clean::TraitWithExtraInfo>>>,
     /// Used while populating `external_traits` to ensure we don't process the same trait twice at
     /// the same time.
-    crate active_extern_traits: FxHashSet<DefId>,
+    crate active_extern_traits: FxHashSet<RealDefId>,
     // The current set of type and lifetime substitutions,
     // for expanding type aliases at the HIR level:
     /// Table `DefId` of type parameter -> substituted type
-    crate ty_substs: FxHashMap<DefId, clean::Type>,
+    crate ty_substs: FxHashMap<RealDefId, clean::Type>,
     /// Table `DefId` of lifetime parameter -> substituted lifetime
-    crate lt_substs: FxHashMap<DefId, clean::Lifetime>,
+    crate lt_substs: FxHashMap<RealDefId, clean::Lifetime>,
     /// Table `DefId` of const parameter -> substituted const
-    crate ct_substs: FxHashMap<DefId, clean::Constant>,
+    crate ct_substs: FxHashMap<RealDefId, clean::Constant>,
     /// Table synthetic type parameter for `impl Trait` in argument position -> bounds
     crate impl_trait_bounds: FxHashMap<ImplTraitParam, Vec<clean::GenericBound>>,
     /// Auto-trait or blanket impls processed so far, as `(self_ty, trait_def_id)`.
     // FIXME(eddyb) make this a `ty::TraitRef<'tcx>` set.
-    crate generated_synthetics: FxHashSet<(Ty<'tcx>, DefId)>,
-    crate auto_traits: Vec<DefId>,
+    crate generated_synthetics: FxHashSet<(Ty<'tcx>, RealDefId)>,
+    crate auto_traits: Vec<RealDefId>,
```

Comment:
Hmm, are you sure auto traits have real ids? I thought they were the whole reason fake ids exist.

Reply:
At least it didn't panic in the test suite, but it still might be the case. I can switch it to use fake ids if we want to be sure.

Reply:
I'm very confused how this could possibly work, the fake IDs are generated here: https://github.com/rust-lang/rust/pull/84707/files#diff-89766a1eb13290d035d1d0f4d720ad83bde30ee2dce910559fee69578aca32b4R116. Do you have any idea why this isn't panicking?

Reply:
What I get from the code is, that the `auto_traits` field is never modified, but only used to cache all auto traits that `TyCtx` returns. 

The only usage of the field is here:
https://github.com/rust-lang/rust/blob/d3274e31fc127a7fed4dff7b2dc652ed63f84523/src/librustdoc/clean/auto_trait.rs#L139

I verified it by removing the line and the field and it compiles fine. So you could even replace the `Vec` with a boxed slice.

Reply:
Got it, thanks :) do you want to make a follow up pr changing this to a boxed slice?

Reply:
Sure, I can do that!